daprdocs/content/en/operations/components/pluggable-components/pluggable-components-overview.md

```diff
@@ -0,0 +1,122 @@
+
+As an example, below is a snippet from the protocol buffer definition file service that defines the gRPC service definition for pluggable component state stores ([state.proto]).
+
+```protobuf
+// StateStore service provides a gRPC interface for state store components.
+service StateStore {
+  // Initializes the state store component with the given metadata.
+  rpc Init(InitRequest) returns (InitResponse) {}
+  // Returns a list of implemented state store features.
+  rpc Features(FeaturesRequest) returns (FeaturesResponse) {}
+  // Deletes the specified key from the state store.
+  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+  // Get data from the given key.
+  rpc Get(GetRequest) returns (GetResponse) {}
+  // Sets the value of the specified key.
+  rpc Set(SetRequest) returns (SetResponse) {}
+  // Ping the state store. Used for liveness purposes.
+  rpc Ping(PingRequest) returns (PingResponse) {}
+
+  // Deletes many keys at once.
+  rpc BulkDelete(BulkDeleteRequest) returns (BulkDeleteResponse) {}
+  // Retrieves many keys at once.
+  rpc BulkGet(BulkGetRequest) returns (BulkGetResponse) {}
+  // Set the value of many keys at once.
+  rpc BulkSet(BulkSetRequest) returns (BulkSetResponse) {}
+}
+```
+
+The interface for the `StateStore` service exposes 9 methods:
+
+- 2 methods for initialization and components capability advertisement (Init and Features)
+- 1 method for health-ness or liveness check (Ping)
+- 3 methods for CRUD (Get, Set, Delete)
+- 3 methods for bulk CRUD operations (BulkGet, BulkSet, BulkDelete)
+
+3. **Define the service.** Provide a concrete implementation of the desired service.
+
+As a first step, [protocol buffers](https://developers.google.com/protocol-buffers/docs/overview) and [gRPC](https://grpc.io/docs/what-is-grpc/introduction/) tools are used to create the server code for this service. After that, the next step is to define concrete implementations for these 9 methods.
+
+Each component has a gRPC service definition for its core functionality which is the same as the core component interface. For example:
+
+- **State stores**
+A pluggable state store **must** provide an implementation of the `StateStore` service interface. In addition to this core functionality, some components might also expose functionality under other **optional** services. For example, you can add extra functionality by defining the implementation for a `QueriableStateStore` service and a `TransactionalStateStore` service.
+
+- **Pub/sub**
+ Pluggable pub/sub components only have a single core service interface defined ([pubsub.proto]). They have no optional service interfaces.
+- **Bindings**
+ Pluggable input and output bindings have a single core service definition on [bindings.proto]. They have no optional service interfaces.
```

Comment:
My suggestion for the code change.

Mostly I
* Corrected the header level to 2
* kept the proto buffer sample to sub-item 1
* added a bit more detail to section 2
* provided a more concrete translation of what step 3 means by linking it back to the snippet from step 1. 

```suggestion
## Implementing a pluggable component

In order to implement a pluggable component you need to implement a gRPC service in the component. Implementing the gRPC service requires three steps:

1. **Find the proto definition file.** Proto definitions are provided for each supported service interface (state store, pub/sub, bindings).

Currently, the following component APIs are supported:

- State stores
- Pub/sub
- Bindings


|  Component  |    Type    | gRPC definition  |                       Built-in Reference Implementation                        | Docs                                                                                                                                                                  |
| :---------: | :--------: | :--------------: | :----------------------------------------------------------------------------: | --------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| State Store |  `state`   |  [state.proto]   |  [Redis](https://github.com/dapr/components-contrib/tree/master/state/redis)   | [concept]({{<ref "state-management-overview">}}), [howto]({{<ref "howto-get-save-state">}}), [api spec]({{<ref "state_api">}})                                        |
|   Pub/sub   |  `pubsub`  |  [pubsub.proto]  |  [Redis](https://github.com/dapr/components-contrib/tree/master/pubsub/redis)  | [concept]({{<ref "pubsub-overview">}}), [howto]({{<ref "howto-publish-subscribe">}}), [api spec]({{<ref "pubsub_api">}})                                              |
|  Bindings   | `bindings` | [bindings.proto] | [Kafka](https://github.com/dapr/components-contrib/tree/master/bindings/kafka) | [concept]({{<ref "bindings-overview">}}), [input howto]({{<ref "howto-triggers">}}), [output howto]({{<ref "howto-bindings">}}), [api spec]({{<ref "bindings_api">}}) |

Here's a snippet of the gRPC service definition for pluggable component state stores ([state.proto]).

```protobuf
// StateStore service provides a gRPC interface for state store components.
service StateStore {
  // Initializes the state store component with the given metadata.
  rpc Init(InitRequest) returns (InitResponse) {}
  // Returns a list of implemented state store features.
  rpc Features(FeaturesRequest) returns (FeaturesResponse) {}
  // Ping the state store. Used for liveness purposes.
  rpc Ping(PingRequest) returns (PingResponse) {}

  // Deletes the specified key from the state store.
  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
  // Get data from the given key.
  rpc Get(GetRequest) returns (GetResponse) {}
  // Sets the value of the specified key.
  rpc Set(SetRequest) returns (SetResponse) {}

  // Deletes many keys at once.
  rpc BulkDelete(BulkDeleteRequest) returns (BulkDeleteResponse) {}
  // Retrieves many keys at once.
  rpc BulkGet(BulkGetRequest) returns (BulkGetResponse) {}
  // Set the value of many keys at once.
  rpc BulkSet(BulkSetRequest) returns (BulkSetResponse) {}
}
```

The interface for the `StateStore` service exposes 9 methods:

- 2 methods for initialization and components capability advertisement (Init and Features)
- 1 method for health-ness or liveness check (Ping)
- 3 methods for CRUD (Get, Set, Delete)
- 3 methods for bulk CRUD operations (BulkGet, BulkSet, BulkDelete)

2. **Create service scaffolding.** Use [protocol buffers and gRPC tools](https://grpc.io) to create the necessary scaffolding for the service. You may want to get acquainted with [the gRPC concepts documentation](https://grpc.io/docs/what-is-grpc/core-concepts/).


The tools can generated code targeting [any gRPC-supported language](https://grpc.io/docs/what-is-grpc/introduction/#protocol-buffer-versions). This code will serve as the base for your server and it provide functionality to handle client calls along with infrastructure to decode incoming requests, execute service methods, and encode service responses.

The generated code will is not complete. It will be missing definitions on how to handle Unix Socket Domain integration, integration with your downstream services and, more importantly, concrete code for the methods your target service define.


3. **Define the service.** Provide a concrete implementation of the desired service.

Each component has a service definition for its core functionality. For example:

- **State stores**
A pluggable state store **must** provide an implementation of the `StateStore` service interface. In addition to this core functionality, some components might also expose functionality under other **optional** services. For example, you can add extra functionality by defining the implementation for a `QueriableStateStore` service and a `TransactionalStateStore` service.

- **Pub/sub**
 Pluggable pub/sub components only have a single core service interface defined ([pubsub.proto]). They have no optional service interfaces.
- **Bindings**
 Pluggable input and output bindings have a single core service definition on [bindings.proto]. They have no optional service interfaces.


Following the State Store example from step 1, after using generating its service scaffolding code using gRPC and protocol buffers tools (step 2), 
the next step is to define concrete implementations for the 9 methods defined under `service StateStore`, along with code to initialize and communicate with your dependencies.
That concrete implementation and auxiliary code are the core of your pluggable component: they define how your component behaves when handling gRPC requests from Dapr.
```